drivers/entropy/entropy_nrf5.c

```diff
@@ -8,92 +8,253 @@
 #include <entropy.h>
 #include <atomic.h>
 #include <nrf_rng.h>
+#include <soc.h>
+
+struct rand {
+	u8_t count;
+	u8_t threshold;
+	u8_t first;
+	u8_t last;
+	u8_t rand[0];
+};
+
+#define RAND_DEFINE(name, len) u8_t name[sizeof(struct rand) + len] __aligned(4)
+
+#define RAND_THREAD_LEN (CONFIG_ENTROPY_NRF5_THR_BUF_LEN + 1)
+#define RAND_ISR_LEN    (CONFIG_ENTROPY_NRF5_ISR_BUF_LEN + 1)
 
 struct entropy_nrf5_dev_data {
-	atomic_t user_count;
+	struct k_sem sem_lock;
+	struct k_sem sem_sync;
+
+	RAND_DEFINE(thr, RAND_THREAD_LEN);
+	RAND_DEFINE(isr, RAND_ISR_LEN);
 };
 
 #define DEV_DATA(dev) \
 	((struct entropy_nrf5_dev_data *)(dev)->driver_data)
 
-static inline u8_t entropy_nrf5_get_u8(void)
+static size_t get(struct rand *rng, size_t octets, u8_t *rand)
```

Comment:
It is possible to reorganize get algorithm  to two if statement:
```if (last < first) {
  //  fetch bytes from upper rng buffer
  // update first, last, octets
}

if (octets && (first <= last)
{
  //  fetch bytes from lower rng buffer
}
```

I don't see  a big sense in do so - so I just I signal this.

Reply:
You suggestion translates to:
1 compare and 1 branch
plus
1 branch, 1 compare and 1 branch

whereas my top level if-clause translates to 1 compare and 1 branch (my understanding).

Reply:
Yes, something like this. Code would be more compact, but  cons is increased execution time.
so keep as is.